File: android/native/emarsys/src/main/java/com/emarsys/maui/DotnetEmarsysInbox.java
package com.emarsys.maui;

import androidx.annotation.NonNull;
import com.emarsys.Emarsys;
import com.emarsys.maui.mapper.MessageMapper;
import java.util.List;
import java.util.Map;

public class DotnetEmarsysInbox {

    public interface FetchMessagesResultCallback {
        void onResult(List<Map<String, Object>> messages, Throwable error);
    }

    public static void fetchMessages(@NonNull FetchMessagesResultCallback resultCallback) {
        Emarsys.getMessageInbox().fetchMessages(result -> {
            if (result.getErrorCause() != null) {
                resultCallback.onResult(null, result.getErrorCause());
            } else {
                if (result.getResult() != null) {
                    List<Map<String, Object>> mappedMessages = MessageMapper.map(result.getResult().getMessages());
                    resultCallback.onResult(mappedMessages, null);
                } else {
                    resultCallback.onResult(List.of(), null);
                }
            }
        });
    }

    public static void addTag(@NonNull String tag, @NonNull String messageId, @NonNull CompletionListener completionListener) {
        Emarsys.getMessageInbox().addTag(tag, messageId, completionListener::onCompleted);
    }

    public static void removeTag(@NonNull String tag, @NonNull String messageId, @NonNull CompletionListener completionListener) {
        Emarsys.getMessageInbox().removeTag(tag, messageId, completionListener::onCompleted);
    }

}

File: ios/native/MauiEmarsys/DotnetEmarsysInbox.swift
//
//  Copyright © 2024 Emarsys. All rights reserved.
//

import Foundation
import EmarsysSDK

@objc(DotnetEmarsysInbox)
public class DotnetEmarsysInbox: NSObject {
    
    @objc
    public static func fetchMessages(_ resultCallback: @escaping ([[String: Any]]?, Error?) -> Void) {
        Emarsys.messageInbox.fetchMessages(resultBlock: { result, error in
            if error != nil {
                resultCallback(nil, error)
            } else {
                if let messages = result?.messages {
                    let mappedMessages = MessageMapper.mapMessages(messages)
                    resultCallback(mappedMessages, nil)
                } else {
                    resultCallback([], nil)
                }
            }
        })
    }
    
    @objc
    public static func addTag(_ tag: String, messageId: String, _ completionBlock: @escaping CompletionBlock) {
        Emarsys.messageInbox.addTag(tag: tag, messageId: messageId, completionBlock: completionBlock)
    }
    
    @objc
    public static func removeTag(_ tag: String, messageId: String, _ completionBlock: @escaping CompletionBlock) {
        Emarsys.messageInbox.removeTag(tag: tag, messageId: messageId, completionBlock: completionBlock)
    }
}
